src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -0,0 +1,754 @@
+    public int Length => 1;
+    public object this[int i] => null;
+    public static void Main()
+    {
+        System.Runtime.CompilerServices.ITuple t = new C();
+        if (t.Length != 1) throw null;
+        if (t[0] != null) throw null;
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: "");
+        }
+
+        [Fact]
+        public void ITupleFromObject()
+        {
+            // - should match when input type is object
+            var source =
+@"using System;
+using System.Runtime.CompilerServices;
+public class C : ITuple
+{
+    int ITuple.Length => 3;
+    object ITuple.this[int i] => i + 3;
+    public static void Main()
+    {
+        object t = new C();
+        Console.WriteLine(t is (3, 4)); // false
+        Console.WriteLine(t is (3, 4, 5)); // TRUE
+        Console.WriteLine(t is (3, 0, 5)); // false
+        Console.WriteLine(t is (3, 4, 5, 6)); // false
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics();
+            var expectedOutput =
+@"False
+True
+False
+False";
+            var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact]
+        public void ITupleMissing()
+        {
+            // - should not match when ITuple is missing
+            var source =
+@"using System;
+public class C
+{
+    public static void Main()
+    {
+        object t = new C();
+        Console.WriteLine(t is (3, 4, 5));
+    }
+}
+";
+            // Use a version of the platform APIs that lack ITuple
+            var compilation = CreateCompilationWithMscorlib40AndSystemCore(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics(
+                // (7,32): error CS1061: 'object' does not contain a definition for 'Deconstruct' and no accessible extension method 'Deconstruct' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)
+                //         Console.WriteLine(t is (3, 4, 5));
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "(3, 4, 5)").WithArguments("object", "Deconstruct").WithLocation(7, 32),
+                // (7,32): error CS8129: No suitable Deconstruct instance or extension method was found for type 'object', with 3 out parameters and a void return type.
+                //         Console.WriteLine(t is (3, 4, 5));
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "(3, 4, 5)").WithArguments("object", "3").WithLocation(7, 32)
+                );
+        }
+
+        [Fact]
+        public void ITupleIsClass()
+        {
+            // - should not match when ITuple is a class
+            var source =
+@"using System;
+namespace System.Runtime.CompilerServices
+{
+    public class ITuple
+    {
+        int Length => 3;
```

Comment:
>int [](start = 8, length = 3)

Consider making these properties `public` so it's more obvious the failure is due to the `class` rather than inaccessible properties.